### PR TITLE
Add the option to configure a chip check mark color

### PR DESCRIPTION
**Pull request by clocksmith**
## Description

* Adds checkmarkColor to the ChipTheme
* Adds an interface for CheckmarkableChip (for FilterChip and InputChip) with checkmarkColor
* Move showCheckmark into CheckmarkableChip

## Related Issues

https://github.com/flutter/flutter/issues/29626

## Tests

I added the following tests:

chip_theme_test
 * added checkmarkColor to lerp tests
chip_test
 * Added tests for checkmark color combinations for InputChip and FilterChip

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/chip.dart [hunk 1]
```diff
@@ -245,6 +245,38 @@ abstract class DeletableChipAttributes {
...
+/// provide or enforce them).
+///
+/// See also:
+///
+///  * [InputChip], a chip that represents a complex piece of information, such
+///    as an entity (person, place, or thing) or conversational text, in a
+///    compact form.
+///  * [FilterChip], uses tags or descriptive words as a way to filter content.
+///  * <https://material.io/design/components/chips.html>
+abstract class CheckmarkableChipAttributes {
```
Comment (johnsonmh) [inline packages/flutter/lib/src/material/chip.dart]: I know there's already a pattern of `FooChipAttributes`, but my personal preference would be to just add these parameters individually on any "CheckMarkable" chip. Then the chip types can pipe the params to `RawMaterialChip`. 

I guess there isn't really a con to having this though. Let's see what others think!

File: packages/flutter/lib/src/material/chip.dart [hunk 2]
```diff
@@ -245,6 +245,38 @@ abstract class DeletableChipAttributes {
...
+  // This class is intended to be used as an interface, and should not be
+  // extended directly.
+  factory CheckmarkableChipAttributes._() => null;
+
+  /// Whether or not to show a check mark when [selected] is true.
+  ///
+  /// Defaults to true.
+  final bool showCheckmark;
```
Comment (johnsonmh) [inline packages/flutter/lib/src/material/chip.dart]: To stick with the pattern, this should probably be a getter.

```suggestion
  bool get showCheckmark;
```

File: packages/flutter/lib/src/material/chip.dart [hunk 3]
```diff
@@ -245,6 +245,38 @@ abstract class DeletableChipAttributes {
...
+
+  /// Color of the chip's check mark when a check mark is visible.
```
Comment (johnsonmh) [inline packages/flutter/lib/src/material/chip.dart]:
```suggestion
  /// [Color] of the chip's check mark when a check mark is visible.
```

File: packages/flutter/lib/src/material/chip.dart [hunk 4]
```diff
@@ -1719,6 +1767,8 @@ class _RawChipState extends State<RawChip> with TickerProviderStateMixin<RawChip
     final double pressElevation = widget.pressElevation ?? chipTheme.pressElevation ?? _defaultPressElevation;
     final Color shadowColor = widget.shadowColor ?? chipTheme.shadowColor ?? _defaultShadowColor;
     final Color selectedShadowColor = widget.selectedShadowColor ?? chipTheme.selectedShadowColor ?? _defaultShadowColor;
+    final Color checkmarkColor = widget.checkmarkColor ?? chipTheme.checkmarkColor;
+    final bool showCheckmark = widget.showCheckmark ?? chipTheme.showCheckmark ?? true;
```
Comment (johnsonmh) [inline packages/flutter/lib/src/material/chip.dart]: Was the implicit behavior before that the checkmark would show if not specified? I feel like I would expect `showCheckmark` to default to false.

File: packages/flutter/lib/src/material/chip.dart [hunk 5]
```diff
@@ -2584,13 +2639,19 @@ class _RenderChip extends RenderBox {
...
-      case Brightness.dark:
-        paintColor = theme.showAvatar ? Colors.black : Colors.white.withAlpha(_kCheckmarkAlpha);
-        break;
+    if (theme.checkmarkColor != null) {
+      paintColor = theme.checkmarkColor;
+    } else {
+      switch (theme.brightness) {
+        case Brightness.light:
+          paintColor = theme.showAvatar ? Colors.white : Colors.black.withAlpha(
+              _kCheckmarkAlpha);
```
Comment (johnsonmh) [inline packages/flutter/lib/src/material/chip.dart]: nit:

this should all be one line.

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#if-you-have-a-newline-after-some-opening-punctuation-match-it-on-the-closing-punctuation

File: packages/flutter/lib/src/material/chip.dart [hunk 6]
```diff
@@ -2584,13 +2639,19 @@ class _RenderChip extends RenderBox {
...
+          break;
+        case Brightness.dark:
+          paintColor = theme.showAvatar ? Colors.black : Colors.white.withAlpha(
+              _kCheckmarkAlpha);
```
Comment (johnsonmh) [inline packages/flutter/lib/src/material/chip.dart]: same here https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#if-you-have-a-newline-after-some-opening-punctuation-match-it-on-the-closing-punctuation

File: packages/flutter/lib/src/material/chip_theme.dart [hunk 7]
```diff
@@ -326,6 +328,19 @@ class ChipThemeData extends Diagnosticable {
   ///  * [shadowColor]
   final Color selectedShadowColor;
 
+  /// Whether or not to show a check mark when [selected] is true.
+  ///
+  /// For instance, the [ChoiceChip] sets this to false so that it can be
+  /// be selected without showing the check mark.
+  ///
+  /// Defaults to true.
+  final bool showCheckmark;
```
Comment (johnsonmh) [inline packages/flutter/lib/src/material/chip_theme.dart]: Ah I think I misunderstood above (when I said this should default to true).

Would you consider changing the name to `showCheckmarkWhenSelected`? Or is that too wordy?

File: packages/flutter/test/material/chip_test.dart [hunk 8]
```diff
@@ -130,6 +132,66 @@ Future<void> _testConstrainedLabel(
   expect(chipSize.height, chipParentHeight);
 }
 
+Future<void> _pumpCheckmarkChip<T>(
```
Comment (johnsonmh) [inline packages/flutter/test/material/chip_test.dart]: The indentation here should be:

```dart
Future<void> _pumpCheckmarkChip<T>(
  WidgetTester tester, {
  Color themeColor,
  Color color,
  Brightness brightness = Brightness.light
}) async {
```

File: packages/flutter/test/material/chip_test.dart [hunk 9]
```diff
@@ -130,6 +132,66 @@ Future<void> _testConstrainedLabel(
...
+      showCheckmark: true,
+      checkmarkColor: color,
+      onSelected: (bool _) { },
+    );
+  } else {
+    throw UnsupportedError('Chip type must be InputChip of FilterChip');
+  }
+
+  await tester.pumpWidget(
+      _wrapForChip(
```
Comment (johnsonmh) [inline packages/flutter/test/material/chip_test.dart]: two spaces of indentation here.

File: packages/flutter/test/material/chip_test.dart [hunk 10]
```diff
@@ -130,6 +132,66 @@ Future<void> _testConstrainedLabel(
...
+  } else if (T == FilterChip) {
+    chip = FilterChip(
+      label: const Text('FilterChip'),
+      selected: true,
+      showCheckmark: true,
+      checkmarkColor: color,
+      onSelected: (bool _) { },
+    );
+  } else {
+    throw UnsupportedError('Chip type must be InputChip of FilterChip');
```
Comment (johnsonmh) [inline packages/flutter/test/material/chip_test.dart]: Rather than checking the type of the generic `T`, consider having this function take a `Widget chip` parameter, then make separate functions for creating the chip types.

This would keep you from having to 1) check the type and 2) throw an error when there is an "unsupported" chip type. 

Calling this function would then look something like:

```dart
_pumpCheckmarkChip(tester, themeColor: Color(...), chip: _filterChip(checkMarkColor: Color(...)));
```

File: packages/flutter/test/material/chip_test.dart [hunk 11]
```diff
@@ -130,6 +132,66 @@ Future<void> _testConstrainedLabel(
...
+                ),
+                child: chip,
+              );
+            },
+          )
+      )
+  );
+}
+
+void _expectCheckmarkColor(Finder finder, Color color, { bool isDark = false }) {
```
Comment (johnsonmh) [inline packages/flutter/test/material/chip_test.dart]: This function actually does more than check the checkmark color.

Consider making two functions.

```dart
_expectCheckmarkColor(Finder finer, color)
// and
_expectChipOverlay(Finder finder, { bool isDark = false });
```

And then call those both explicitly in each test.

File: packages/flutter/lib/src/material/chip.dart (hunk 4 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/chip.dart]: The default behavior is to show the check mark when the chip is selected

File: packages/flutter/lib/src/material/chip_theme.dart (hunk 7 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/chip_theme.dart]: It already existed as showCheckmark, I just moved it to Checkmarkable interface

File: packages/flutter/lib/src/material/chip.dart (hunk 2 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/chip.dart]: Done

File: packages/flutter/lib/src/material/chip.dart (hunk 3 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/chip.dart]: Done

File: packages/flutter/lib/src/material/chip.dart (hunk 1 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/chip.dart]: I did this mainly for consistency.  A new interface had to be added because although ChoiceChip, InputChip, and FilterChip are all selectable, ChoiceChip is not checkmarkable.

It is nice to use the interfaces because the documentation for each filed is inherently centralized (of course macros could do this too)

File: packages/flutter/test/material/chip_test.dart (hunk 8 above)
Comment (clocksmith) [reply packages/flutter/test/material/chip_test.dart]: Done

File: packages/flutter/test/material/chip_test.dart (hunk 10 above)
Comment (clocksmith) [reply packages/flutter/test/material/chip_test.dart]: Good idea! done

File: packages/flutter/test/material/chip_test.dart (hunk 11 above)
Comment (clocksmith) [reply packages/flutter/test/material/chip_test.dart]: It checks the overlay first because the expect paints API has to cover all the paths that are painted in order. The check mark is the second path to be painted, so we have to add the first path. However, I was able to remove the color from the first path, since that is not what we care about, and it allowed for isDark to be removed.

File: packages/flutter/test/material/chip_test.dart [hunk 12]
```diff
@@ -130,6 +132,64 @@ Future<void> _testConstrainedLabel(
...
+  );
+}
+
+void _expectCheckmarkColor(Finder finder, Color color) {
+  expect(
+    finder,
+    paints
+      // The first path that is painted is the selection overlay. We do not care
+      // how it is painted but it has to be added it to this pattern so that the
+      // check mark can be checked next.
```
Comment (johnsonmh) [inline packages/flutter/test/material/chip_test.dart]: Ohhh I see now, this comment is super helpful, thanks!

File: packages/flutter/test/material/chip_test.dart [hunk 13]
```diff
@@ -130,6 +132,64 @@ Future<void> _testConstrainedLabel(
...
+      ..path()
+      // The second path that is painted is the check mark.
+      ..path(
```
Comment (johnsonmh) [inline packages/flutter/test/material/chip_test.dart]: nit: this could probably be on one line:

```dart
...path(color: color),
```

File: packages/flutter/test/material/chip_test.dart [hunk 14]
```diff
@@ -130,6 +132,64 @@ Future<void> _testConstrainedLabel(
   expect(chipSize.height, chipParentHeight);
 }
 
+Widget _selectedInputChip({Color checkmarkColor}) {
```
Comment (johnsonmh) [inline packages/flutter/test/material/chip_test.dart]: Nit spaces next to curly braces (here and on the function below).

```suggestion
Widget _selectedInputChip({ Color checkmarkColor }) {
```

**Review by johnsonmh — APPROVED** [review]
LGTM!

File: packages/flutter/test/material/chip_test.dart (hunk 13 above)
Comment (clocksmith) [reply packages/flutter/test/material/chip_test.dart]: Done

File: packages/flutter/test/material/chip_test.dart (hunk 14 above)
Comment (clocksmith) [reply packages/flutter/test/material/chip_test.dart]: Done

**Review by johnsonmh — APPROVED** [review]

File: packages/flutter/lib/src/material/chip_theme.dart [hunk 15]
```diff
@@ -326,6 +328,19 @@ class ChipThemeData extends Diagnosticable {
   ///  * [shadowColor]
   final Color selectedShadowColor;
 
+  /// Whether or not to show a check mark when [selected] is true.
+  ///
+  /// For instance, the [ChoiceChip] sets this to false so that it can be
+  /// be selected without showing the check mark.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip_theme.dart]: be be

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/chip_theme.dart (hunk 7 above)
Comment (rami-a) [inline packages/flutter/lib/src/material/chip_theme.dart]: I don't think you should say `defaults to true` here since in the context of the ThemeData it defaults to null. I'd just remove the line

**Review by rami-a — APPROVED** [review]

File: packages/flutter/lib/src/material/chip_theme.dart (hunk 7 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/chip_theme.dart]: keeping for file consistency
